File: frontend/src/utils/recurrence.test.ts
import { describe, expect, it } from "vitest";
import {
  describeRecurrence,
  dueDatesInRange,
  isChoreDueOn,
  isChoreDueToday,
  parseDueDate,
  parseRecurrenceDays,
  serializeRecurrenceDays,
  toDateInputValue,
  type RecurrenceConfig,
} from "./recurrence";

// Mirrors backend/internal/services/recurrence_test.go — this is a
// faithful JS port of that Go logic, so the test cases (and the Monday
// anchor date) match 1:1.

function date(year: number, month: number, day: number): Date {
  return new Date(year, month - 1, day);
}

describe("parseRecurrenceDays", () => {
  it.each([
    ["nil", undefined, []],
    ["empty string", "", []],
    ["null", null, []],
    ["valid", "[1,3,5]", [1, 3, 5]],
    ["out of range dropped", "[-1,0,6,7]", [0, 6]],
    ["malformed json", "not json", []],
    ["not an array", '{"a":1}', []],
  ] as const)("%s", (_name, raw, want) => {
    expect(parseRecurrenceDays(raw)).toEqual(want);
  });
});

describe("serializeRecurrenceDays", () => {
  it("empty", () => {
    expect(serializeRecurrenceDays([])).toBe("[]");
  });
  it("dedupes and sorts", () => {
    expect(serializeRecurrenceDays([5, 1, 1, 3])).toBe("[1,3,5]");
  });
  it("single", () => {
    expect(serializeRecurrenceDays([2])).toBe("[2]");
  });
});

describe("isChoreDueOn", () => {
  const anchor = date(2026, 8, 17); // a Monday

  const cases: [string, RecurrenceConfig, Date, boolean][] = [
    ["none: exact due date", { recurrence: "none", dueDate: anchor, recurrenceDays: null }, anchor, true],
    [
      "none: different date",
      { recurrence: "none", dueDate: anchor, recurrenceDays: null },
      date(2026, 8, 18),
      false,
    ],
    ["daily: on anchor", { recurrence: "daily", dueDate: anchor, recurrenceDays: null }, anchor, true],
    [
      "daily: well after anchor",
      { recurrence: "daily", dueDate: anchor, recurrenceDays: null },
      date(2026, 9, 16),
      true,
    ],
    [
      "daily: before anchor never due",
      { recurrence: "daily", dueDate: anchor, recurrenceDays: null },
      date(2026, 8, 16),
      false,
    ],
    [
      "weekly: same weekday next week",
      { recurrence: "weekly", dueDate: anchor, recurrenceDays: null },
      date(2026, 8, 24),
      true,
    ],
    [
      "weekly: different weekday",
      { recurrence: "weekly", dueDate: anchor, recurrenceDays: null },
      date(2026, 8, 18),
      false,
    ],
    ["weekdays: Monday", { recurrence: "weekdays", dueDate: anchor, recurrenceDays: null }, date(2026, 8, 17), true],
    [
      "weekdays: Saturday excluded",
      { recurrence: "weekdays", dueDate: anchor, recurrenceDays: null },
      date(2026, 8, 22),
      false,
    ],
    [
      "weekdays: Sunday excluded",
      { recurrence: "weekdays", dueDate: anchor, recurrenceDays: null },
      date(2026, 8, 23),
      false,
    ],
    [
      "custom: matches one of the configured days",
      { recurrence: "custom", dueDate: anchor, recurrenceDays: "[1,3]" }, // Mon, Wed
      date(2026, 8, 19), // Wednesday
      true,
    ],
    [
      "custom: does not match",
      { recurrence: "custom", dueDate: anchor, recurrenceDays: "[1,3]" },
      date(2026, 8, 20), // Thursday
      false,
    ],
    [
      "custom: no days configured never due",
      { recurrence: "custom", dueDate: anchor, recurrenceDays: null },
      anchor,
      false,
    ],
  ];

  it.each(cases)("%s", (_name, config, target, want) => {
    expect(isChoreDueOn(config, target)).toBe(want);
  });
});

describe("isChoreDueToday", () => {
  it("a daily chore anchored well in the past is always due today", () => {
    const config: RecurrenceConfig = { recurrence: "daily", dueDate: date(2000, 1, 1), recurrenceDays: null };
    expect(isChoreDueToday(config)).toBe(true);
  });
});

describe("describeRecurrence", () => {
  const anchor = date(2026, 8, 17); // Monday

  it("daily", () => {
    expect(describeRecurrence({ recurrence: "daily", dueDate: anchor, recurrenceDays: null })).toBe("Daily");
  });
  it("weekly", () => {
    expect(describeRecurrence({ recurrence: "weekly", dueDate: anchor, recurrenceDays: null })).toBe(
      "Weekly (Mon)",
    );
  });
  it("weekdays", () => {
    expect(describeRecurrence({ recurrence: "weekdays", dueDate: anchor, recurrenceDays: null })).toBe("Weekdays");
  });
  it("custom with days", () => {
    expect(describeRecurrence({ recurrence: "custom", dueDate: anchor, recurrenceDays: "[1,3,5]" })).toBe(
      "Custom (Mon, Wed, Fri)",
    );
  });
});

describe("dueDatesInRange", () => {
  const anchor = date(2026, 8, 17); // Monday

  it("weekly within a two week range hits exactly twice", () => {
    const got = dueDatesInRange(
      { recurrence: "weekly", dueDate: anchor, recurrenceDays: null },
      anchor,
      date(2026, 8, 30),
    );
    expect(got.map((d) => d.getTime())).toEqual([anchor.getTime(), date(2026, 8, 24).getTime()]);
  });

  it("range entirely before anchor returns nothing", () => {
    const got = dueDatesInRange(
      { recurrence: "daily", dueDate: anchor, recurrenceDays: null },
      date(2026, 8, 7),
      date(2026, 8, 16),
    );
    expect(got).toEqual([]);
  });

  it("one-time chore returns at most one date", () => {
    const got = dueDatesInRange(
      { recurrence: "none", dueDate: anchor, recurrenceDays: null },
      date(2026, 8, 12),
      date(2026, 8, 22),
    );
    expect(got).toHaveLength(1);
    expect(got[0].getTime()).toBe(anchor.getTime());
  });
});

describe("parseDueDate", () => {
  it("parses a YYYY-MM-DD date string as a local calendar date", () => {
    const parsed = parseDueDate("2026-08-17");
    expect(parsed.getFullYear()).toBe(2026);
    expect(parsed.getMonth()).toBe(7); // 0-indexed
    expect(parsed.getDate()).toBe(17);
  });

  it("truncates a full ISO timestamp to just the date part", () => {
    const parsed = parseDueDate("2026-08-17T15:30:00.000Z");
    expect(parsed.getFullYear()).toBe(2026);
    expect(parsed.getMonth()).toBe(7);
    expect(parsed.getDate()).toBe(17);
  });
});

describe("toDateInputValue", () => {
  it("formats as YYYY-MM-DD with zero-padding", () => {
    expect(toDateInputValue(date(2026, 1, 5))).toBe("2026-01-05");
  });

  it("round-trips with parseDueDate", () => {
    const original = "2026-08-17";
    expect(toDateInputValue(parseDueDate(original))).toBe(original);
  });
});

File: backend/internal/services/recurrence_test.go
package services

import (
	"testing"
	"time"
)

func date(y int, m time.Month, d int) time.Time {
	return time.Date(y, m, d, 0, 0, 0, 0, time.UTC)
}

func TestParseRecurrenceDays(t *testing.T) {
	tests := []struct {
		name string
		raw  *string
		want []int
	}{
		{"nil", nil, []int{}},
		{"empty string", ptr(""), []int{}},
		{"valid", ptr(`[1,3,5]`), []int{1, 3, 5}},
		{"out of range dropped", ptr(`[-1,0,6,7]`), []int{0, 6}},
		{"malformed json", ptr(`not json`), []int{}},
		{"not an array", ptr(`{"a":1}`), []int{}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := ParseRecurrenceDays(tt.raw)
			if !intSliceEqual(got, tt.want) {
				t.Errorf("ParseRecurrenceDays(%v) = %v, want %v", derefOrNil(tt.raw), got, tt.want)
			}
		})
	}
}

func TestSerializeRecurrenceDays(t *testing.T) {
	tests := []struct {
		name string
		days []int
		want string
	}{
		{"empty", []int{}, "[]"},
		{"dedupes and sorts", []int{5, 1, 1, 3}, "[1,3,5]"},
		{"single", []int{2}, "[2]"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := SerializeRecurrenceDays(tt.days); got != tt.want {
				t.Errorf("SerializeRecurrenceDays(%v) = %q, want %q", tt.days, got, tt.want)
			}
		})
	}
}

func TestIsChoreDueOn(t *testing.T) {
	anchor := date(2026, time.August, 17) // a Monday

	tests := []struct {
		name   string
		config RecurrenceConfig
		target time.Time
		want   bool
	}{
		{"none: exact due date", RecurrenceConfig{Recurrence: "none", DueDate: anchor}, anchor, true},
		{"none: different date", RecurrenceConfig{Recurrence: "none", DueDate: anchor}, anchor.AddDate(0, 0, 1), false},
		{"none: time-of-day ignored", RecurrenceConfig{Recurrence: "none", DueDate: anchor.Add(5 * time.Hour)}, anchor, true},

		{"daily: on anchor", RecurrenceConfig{Recurrence: "daily", DueDate: anchor}, anchor, true},
		{"daily: well after anchor", RecurrenceConfig{Recurrence: "daily", DueDate: anchor}, anchor.AddDate(0, 0, 30), true},
		{"daily: before anchor never due", RecurrenceConfig{Recurrence: "daily", DueDate: anchor}, anchor.AddDate(0, 0, -1), false},

		{"weekly: same weekday next week", RecurrenceConfig{Recurrence: "weekly", DueDate: anchor}, anchor.AddDate(0, 0, 7), true},
		{"weekly: different weekday", RecurrenceConfig{Recurrence: "weekly", DueDate: anchor}, anchor.AddDate(0, 0, 1), false},
		{"weekly: before anchor never due even if same weekday", RecurrenceConfig{Recurrence: "weekly", DueDate: anchor}, anchor.AddDate(0, 0, -7), false},

		{"weekdays: Monday", RecurrenceConfig{Recurrence: "weekdays", DueDate: anchor}, date(2026, time.August, 17), true},
		{"weekdays: Saturday excluded", RecurrenceConfig{Recurrence: "weekdays", DueDate: anchor}, date(2026, time.August, 22), false},
		{"weekdays: Sunday excluded", RecurrenceConfig{Recurrence: "weekdays", DueDate: anchor}, date(2026, time.August, 23), false},
		{"weekdays: Friday included", RecurrenceConfig{Recurrence: "weekdays", DueDate: anchor}, date(2026, time.August, 21), true},

		{
			"custom: matches one of the configured days",
			RecurrenceConfig{Recurrence: "custom", DueDate: anchor, RecurrenceDays: ptr(`[1,3]`)}, // Mon, Wed
			date(2026, time.August, 19),                                                           // Wednesday
			true,
		},
		{
			"custom: does not match",
			RecurrenceConfig{Recurrence: "custom", DueDate: anchor, RecurrenceDays: ptr(`[1,3]`)},
			date(2026, time.August, 20), // Thursday
			false,
		},
		{
			"custom: no days configured never due",
			RecurrenceConfig{Recurrence: "custom", DueDate: anchor, RecurrenceDays: nil},
			anchor,
			false,
		},

		{"unknown recurrence value defaults to never due", RecurrenceConfig{Recurrence: "bogus", DueDate: anchor}, anchor, false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := IsChoreDueOn(tt.config, tt.target); got != tt.want {
				t.Errorf("IsChoreDueOn() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestIsChoreDueToday(t *testing.T) {
	// IsChoreDueToday is IsChoreDueOn(chore, time.Now()) — a daily chore
	// anchored in the past should always be due "today," regardless of
	// when the test actually runs.
	config := RecurrenceConfig{Recurrence: "daily", DueDate: date(2000, time.January, 1)}
	if !IsChoreDueToday(config) {
		t.Error("expected a daily chore anchored well in the past to be due today")
	}
}

func TestDescribeRecurrence(t *testing.T) {
	anchor := date(2026, time.August, 17) // Monday

	tests := []struct {
		name   string
		config RecurrenceConfig
		want   string
	}{
		{"none", RecurrenceConfig{Recurrence: "none", DueDate: anchor}, "One-time (8/17/2026)"},
		{"daily", RecurrenceConfig{Recurrence: "daily", DueDate: anchor}, "Daily"},
		{"weekly", RecurrenceConfig{Recurrence: "weekly", DueDate: anchor}, "Weekly (Mon)"},
		{"weekdays", RecurrenceConfig{Recurrence: "weekdays", DueDate: anchor}, "Weekdays"},
		{"custom with days", RecurrenceConfig{Recurrence: "custom", DueDate: anchor, RecurrenceDays: ptr(`[1,3,5]`)}, "Custom (Mon, Wed, Fri)"},
		{"custom with no days", RecurrenceConfig{Recurrence: "custom", DueDate: anchor}, "Custom ()"},
		{"unknown falls back to raw value", RecurrenceConfig{Recurrence: "bogus", DueDate: anchor}, "bogus"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := DescribeRecurrence(tt.config); got != tt.want {
				t.Errorf("DescribeRecurrence() = %q, want %q", got, tt.want)
			}
		})
	}
}

func TestDueDatesInRange(t *testing.T) {
	anchor := date(2026, time.August, 17) // Monday

	t.Run("weekly within a two week range hits exactly twice", func(t *testing.T) {
		got := DueDatesInRange(
			RecurrenceConfig{Recurrence: "weekly", DueDate: anchor},
			anchor,
			anchor.AddDate(0, 0, 13),
		)
		want := []time.Time{anchor, anchor.AddDate(0, 0, 7)}
		if len(got) != len(want) {
			t.Fatalf("got %d dates, want %d: %v", len(got), len(want), got)
		}
		for i := range want {
			if !got[i].Equal(want[i]) {
				t.Errorf("date[%d] = %v, want %v", i, got[i], want[i])
			}
		}
	})

	t.Run("range entirely before anchor returns nothing", func(t *testing.T) {
		got := DueDatesInRange(
			RecurrenceConfig{Recurrence: "daily", DueDate: anchor},
			anchor.AddDate(0, 0, -10),
			anchor.AddDate(0, 0, -1),
		)
		if len(got) != 0 {
			t.Errorf("expected no due dates before the anchor, got %v", got)
		}
	})

	t.Run("one-time chore returns at most one date", func(t *testing.T) {
		got := DueDatesInRange(
			RecurrenceConfig{Recurrence: "none", DueDate: anchor},
			anchor.AddDate(0, 0, -5),
			anchor.AddDate(0, 0, 5),
		)
		if len(got) != 1 || !got[0].Equal(anchor) {
			t.Errorf("expected exactly [%v], got %v", anchor, got)
		}
	})
}

func ptr(s string) *string { return &s }

func derefOrNil(s *string) any {
	if s == nil {
		return nil
	}
	return *s
}

func intSliceEqual(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}
